### ISSUE TITLE
Add a HingeJoint2D node to complement HingeJoint3D

### ISSUE THREAD

**Issue by cheesymoo** [godotengine/godot-proposals#1033]
**Describe the project you are working on:**
A 2D game featuring vehicles and trailer attachements

**Describe the problem or limitation you are having in your project:**
I would like to attach 2DKinematic or 2DPhysics bodies to each other via a 2D-HingeJoint

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
2D should have a HingeJoint, it is obviously useful as we have a 3D version of it already

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
It will be analogous to the 3DHingeJoint, just for 2D physics

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I have attempted using other joints and scripting work-arounds, but nothing seems to have the desired effect

**Is there a reason why this should be core and not an add-on in the asset library?:**
It brings the 2D joints into parity with the 3D joints

**Comment by pwab** [comment]
As we have the PinJoint2D that only needs the addition of a max and min value of the rotation_degree (relative to a zero axis) what kind of implementation would this proposal mean?

1. Only add such a feature to PinJoint2D
2. Add the feature and rename the PinJoint2D to HingeJoint2D
3. Add an extra HingeJoint2D with the same functionality as a PinJoint2D with the only difference that it is able of maintaining a max and a min value for the relative rotation

@cheesymoo would you specify this more clearly what you have in mind?

**Comment by cheesymoo** [comment]
@pwab Thanks for sharing your workaround! 

As far as your 3 proposed implementations, any of those would satisfy my needs.

> 1. Only add such a feature (max and min value of the rotation_degree) to PinJoint2D
This is great, it satisfies my needs is simplest
> 2. Add the feature and rename the PinJoint2D to HingeJoint2D
This is bad, because it removes existing nodes and could break other people's games
> 3. Add an extra HingeJoint2D with the same functionality as a PinJoint2D with the only difference that it is able of maintaining a max and a min value for the relative rotation
This is the most ideal as it brings the 2D joints closer to parity with the 3D ones

So I am in support of your proposal 1 or 3. If the community/developers are not concerned about having 2D/3D parity between nodes, than I think 1 is the quickest/easiest solution

**Comment by MrZak-dev** [comment]
I really needed this to create a simple Knee animation in a Human Shaped RigidBody2D , we hope it will se some light in the coming Godot version .

**Comment by pwab** [comment]
Hey @MrZak-dev your knee animation should be a perfect fit for the IK-PR by TwistedTwigleg (https://github.com/godotengine/godot/pull/40347). After building his working 2D branch you can simply open some of his samples [here](https://github.com/TwistedTwigleg/Godot_GSOC_2020_2D_Project). He also implemented constraints with viewable debugging features. (TwoBoneIK should be the best for a human knee but you can also try the FABRIK algorithm for the whole lower limb which is basically the same thing I implemented [here](https://github.com/godotengine/godot-demo-projects/issues/211#issuecomment-362838172) a few years ago.)

The `PhysicalBone` could be also another workaround for the problem @cheesymoo was facing. But I still think a 2DHingeJoint or simply the implementation of rotational constraints into the PinJoint2D could be a great benefit when dealing with such objects.

**Comment by vesk4000** [comment]
Yes, please! I really need this. Also it would be awesome if joints could have angle constraints like they do in Unity.

**Comment by Calinou** [comment]
@vesk4000 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by Athakaspen** [comment]
The motor functionality of the 3D HingeJoint would also be very useful in 2D. I'm currently experimenting with a basic multijointed robot arm-esque model, and need a good way to apply torques around those joints.

**Comment by addmix** [comment]
This functionality, and many other Joint2D related PRs would be solved by a 2D equivalent of the `Generic6DOFJoint3D` node.
These issues could all be solved by a generic Joint2D #4224 #86 #3891 #1033 #1423 #3704 (this one is only kinda solved by a generic joint).
